### PR TITLE
fix 1139

### PR DESCRIPTION
**Pull request by soyuka**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1139 
| License       | MIT
| Doc PR        | na

### CODE REVIEW

**Comment by teohhanhui** [comment]
I think what we really need is tracking of "visited" subresources, so that we can know we have already been through the same path before.

**Comment by dunglas** [comment]
Thanks @soyuka!

File: src/Bridge/Symfony/Routing/ApiLoader.php [hunk 1]
```diff
@@ -152,6 +152,12 @@ private function computeSubresourceOperations(RouteCollection $routeCollection,
                 'collection' => $isCollection,
             ];
 
+            $visiting = "$resourceClass$subresource";
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Oops... The visited tracking is not specific enough.

For example, if I have:

Product -> isRelatedTo
and
Product -> isSimilarTo

It only works for the first of its type.

**Comment by teohhanhui** [comment]
Follow-up to https://github.com/api-platform/core/pull/1144#discussion_r120525352

Provided:

```
Product -[isRelatedTo]-> Product
Product -[isSimilarTo]-> Product
Product -[taxons]-> Taxon
Taxon -[children]-> Taxon
Taxon -[parent]-> Taxon
Taxon -[taxonomy]-> Taxonomy
Taxonomy -[taxons]-> Taxon
```

If we traverse using BFS (I'm not sure how we can do this with DFS?), with a visited tracking format of:

```php
// [
//     '$rootResourceClass' => [
//         '$resourceClass' => [
//             '$property $subresource',
//         ],
//     ],
// ]
[
    'Product' => [
        'Product' => [
            'isRelatedTo Product',
            'isSimilarTo Product',
            'taxons Taxon',
        ],
        'Taxon' => [
            'children Taxon',
            'parent Taxon',
            'taxonomy Taxonomy',
        ],
        'Taxonomy' => [
            'taxons Taxon',
        ],
    ],
    'Taxon' => [
        'Taxon' => [
            'children Taxon',
            'parent Taxon',
            'taxonomy Taxonomy',
        ],
        'Taxonomy' => [
            'taxons Taxon',
        ],
    ],
    'Taxonomy' => [
        'Taxonomy' => [
            'taxons Taxon',
        ],
        'Taxon' => [
            'children Taxon',
            'parent Taxon',
            'taxonomy Taxonomy',
        ],
    ],
]
```

Would result in:

-   api_products_is_related_to_get_subresource
    `/products/{id}/is_related_to`
-   api_products_is_similar_to_get_subresource
    `/products/{id}/is_similar_to`
-   api_products_taxons_get_subresource
    `/products/{id}/taxons`
-   api_products_taxons_children_get_subresource
    `/products/{id}/taxons/{taxonId}/children`
-   api_products_taxons_parent_get_subresource
    `/products/{id}/taxons/{taxonId}/parent`
-   api_products_taxons_taxonomy_get_subresource
    `/products/{id}/taxons/{taxonId}/taxonomy`
-   api_products_taxons_taxonomy_taxons_get_subresource
    `/products/{id}/taxons/{taxonId}/taxonomy/taxons`
-   api_taxons_children_get_subresource
    `/taxons/{id}/children`
-   api_taxons_parent_get_subresource
    `/taxons/{id}/parent`
-   api_taxons_taxonomy_get_subresource
    `/taxons/{id}/taxonomy`
-   api_taxons_taxonomy_taxons_get_subresource
    `/taxons/{id}/taxonomy/taxons`
-   api_taxonomy_taxons_get_subresource
    `/taxonomy/{id}/taxons`
-   api_taxonomy_taxons_children_get_subresource
    `/taxonomy/{id}/taxons/{taxonId}/children`
-   api_taxonomy_taxons_parent_get_subresource
    `/taxonomy/{id}/taxons/{taxonId}/parent`
-   api_taxonomy_taxons_taxonomy_get_subresource
    `/taxonomy/{id}/taxons/{taxonId}/taxonomy`

What do you think @soyuka?

File: src/Bridge/Symfony/Routing/ApiLoader.php (hunk 1 above)
Comment (soyuka) [inline src/Bridge/Symfony/Routing/ApiLoader.php]: Okay I need to add the property then.

**Pull request merged**